File: utils/archive/entry.dart
// Copyright (c) 2012, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

library entry;

import 'dart:io';
import 'archive.dart' as archive;
import 'entry_request.dart';
import 'read_request.dart' as read;
import 'utils.dart';

/**
 * A single file in an archive.
 *
 * This is accessible via [ArchiveInputStream.onEntry].
 */
class ArchiveEntry {
  /**
   * The various properties of this archive entry, as sent over from the C
   * extension.
   */
  final List _properties;

  /**
   * The id of the archive to which this entry belongs. Used to read the entry
   * data. This will be set to null once there's no longer data available to be
   * read for this entry.
   */
  int _archiveId;

  /**
   * The input stream being used to read data from this entry. This is null
   * until [openInputStream] is called.
   */
  InputStream _input;

  // TODO(nweiz): Get rid of this once issue 4202 is fixed.
  /**
   * A future that only exists once [openInputStream] is called, and completes
   * once the input stream is closed.
   *
   * For internal use only.
   */
  Future inputComplete;

  ArchiveEntry.internal(this._properties, this._archiveId) {
    attachFinalizer(this, (id) => call(FREE, id), _id);
  }

  /** Create a new [ArchiveEntry] with default values for all of its fields. */
  static Future<ArchiveEntry> create() {
    return call(NEW).transform((properties) {
      return new archive.ArchiveEntry.internal(properties, null);
    });
  }

  /** The id of the underlying archive entry. */
  int get _id => _properties[0];

  /** If this entry is a hardlink, this is the destination. Otherwise, null. */
  String get hardlink => _properties[1];
  set hardlink(String value) => _set(SET_HARDLINK, 1, value);

  /** The path to this entry in the archive. */
  String get pathname => _properties[2];
  set pathname(String value) => _set(SET_PATHNAME, 2, value);

  /** The path to this entry on disk, */
  String get sourcepath => _properties[3];

  /** If this entry is a symlink, this is the destination. Otherwise, null. */
  String get symlink => _properties[4];
  set symlink(String value) => _set(SET_SYMLINK, 4, value);

  /** The group identifier for this entry. */
  int get gid => _properties[5];
  set gid(int value) => _set(SET_GID, 5, value);

  /** The user identifier for this entry. */
  int get uid => _properties[6];
  set uid(int value) => _set(SET_UID, 6, value);

  /** The permissions bitmask for this entry. */
  int get perm_mask => _properties[7];
  set perm_mask(int value) => _set(SET_PERM, 7, value);

  /**
   * The String representation of the permissions for this entry.
   *
   * Note that if you set [perm_mask], this value will not change.
   */
  String get strmode => _properties[8];

  /** The name of the group this entry belongs to. */
  String get gname => _properties[9];
  set gname(String value) => _set(SET_GNAME, 9, value);

  /** The name of the user this entry belongs to. */
  String get uname => _properties[10];
  set uname(String value) => _set(SET_UNAME, 10, value);

  /**
   * The file flag bits that should be set for this entry.
   *
   * Note that if you set [fflags_text], this value will not change, and vice
   * versa.
   */
  int get fflags_set => _properties[11];
  set fflags_set(int value) => _set(SET_FFLAGS_SET, 11, value);

  /**
   * The file flag bits that should be cleared for this entry.
   *
   * Note that if you set [fflags_text], this value will not change, and vice
   * versa.
   */
  int get fflags_clear => _properties[12];
  set fflags_clear(int value) => _set(SET_FFLAGS_CLEAR, 12, value);

  /**
   * The textual representation of the file flags for this entry.
   *
   * Note that if you set [fflags_set] or [fflags_clear], this value will not
   * change, and vice versa.
   */
  String get fflags_text => _properties[13];

  /** The filetype bitmask for this entry. */
  int get filetype_mask => _properties[14];
  set filetype_mask(int value) => _set(SET_FILETYPE, 14, value);

  /** The filetype and permissions bitmask for this entry. */
  int get mode_mask => _properties[15];
  set mode_mask(int value) => _set(SET_MODE, 15, value);

  /** The size of this entry in bytes, or null if it's unset. */
  int get size => _properties[16];
  set size(int value) => _set(SET_SIZE, 16, value);

  /** The ID of the device containing this entry, or null if it's unset. */
  int get dev => _properties[17];
  set dev(int value) => _set(SET_DEV, 17, value);

  /** The major number of the ID of the device containing this entry. */
  int get devmajor => _properties[18];
  set devmajor(int value) => _set(SET_DEVMAJOR, 18, value);

  /** The minor number of the ID of the device containing this entry. */
  int get devminor => _properties[19];
  set devminor(int value) => _set(SET_DEVMINOR, 19, value);

  /** The inode number of this entry, or null if it's unset. */
  int get ino => _properties[20];
  set ino(int value) => _set(SET_INO, 20, value);

  /** The number of references to this entry. */
  int get nlink => _properties[21];
  set nlink(int value) => _set(SET_NLINK, 21, value);

  /** The device ID of this entry. */
  int get rdev => _properties[22];
  set rdev(int value) => _set(SET_RDEV, 22, value);

  /** The major number of the device ID of this entry. */
  int get rdevmajor => _properties[23];
  set rdevmajor(int value) => _set(SET_RDEVMAJOR, 23, value);

  /** The minor number of the device ID of this entry. */
  int get rdevminor => _properties[24];
  set rdevminor(int value) => _set(SET_RDEVMINOR, 24, value);

  /** The last time this entry was accessed, or null if it's unset. */
  Date get atime => _fromMs(_properties[25]);
  set atime(Date value) => _set(SET_ATIME, 25, _toMs(value));

  /** The time this entry was created, or null if it's unset. */
  Date get birthtime => _fromMs(_properties[26]);
  set birthtime(Date value) => _set(SET_BIRTHTIME, 26, _toMs(value));

  /**
   * The last time an inode property of this entry was changed, or null if it's
   * unset.
   */
  Date get ctime => _fromMs(_properties[27]);
  set ctime(Date value) => _set(SET_CTIME, 27, _toMs(value));

  /** The last time this entry was modified, or null if it's unset. */
  Date get mtime => _fromMs(_properties[28]);
  set mtime(Date value) => _set(SET_MTIME, 28, _toMs(value));

  /** Whether [openInputStream] has been called. */
  bool get isInputOpen => _input != null;

  /** Create a deep copy of this [ArchiveEntry]. */
  Future<ArchiveEntry> clone() {
    return call(CLONE, _id).
      transform((array) => new archive.ArchiveEntry.internal(array, null));
  }

  /**
   * Consumes the entire contents of this entry at once and returns it wrapped
   * in a [CompleteArchiveEntry]. All metadata fields in the returned entry are
   * copies of the fields in this entry.
   *
   * This may not be called if [openInputStream] is called, and vice versa.
   */
  Future<CompleteArchiveEntry> readAll() {
    var stream = openInputStream();
    var buffer = <int>[];
    var completer = new Completer<List<int>>();

    stream.onData = () => buffer.addAll(stream.read());
    stream.onError = completer.completeException;
    stream.onClosed = () => completer.complete(buffer);

    return Futures.wait([call(CLONE, _id), completer.future])
      .transform((list) => new CompleteArchiveEntry._(list[0], list[1]));
  }

  /**
   * Set a property value with index [value] on the local representation of the
   * archive entry and on the native representation.
   */
  void _set(int requestType, int index, value) {
    _properties[index] = value;
    // Since the native code processes messages in order, the SET_* messages
    // will be received and processed before any further messages.
    call(requestType, _id, [value]).then((_) {});
  }

  /**
   * Converts [ms], the (possibly null) number of milliseconds since the epoch
   * into a Date object (which may also be null).
   */
  Date _fromMs(int ms) {
    if (ms == null) return null;
    return new Date.fromMillisecondsSinceEpoch(ms);
  }

  /**
   * Converts [date], which may be null, into the number of milliseconds since
   * the epoch (which may also be null).
   */
  int _toMs(Date date) {
    if (date == null) return null;
    return date.millisecondsSinceEpoch;
  }

  /**
   * Creates a new input stream for reading the contents of this entry.
   *
   * The contents of an entry must be consumed before the next entry is read
   * from the parent [ArchiveInputStream]. This means that [openInputStream]
   * must be called from the [ArchiveInputStream.onEntry] callback, or before
   * the future returned by that callback completes. Once the next entry has
   * been read, calling [openInputStream] will throw an [ArchiveException].
   *
   * Only one input stream may be opened per entry.
   */
  InputStream openInputStream() {
    if (_archiveId == null) {
      throw new UnsupportedError("Cannot open input stream for "
          "archive entry $pathname.");
    } else if (_input != null) {
      throw new UnsupportedError("An input stream has already been"
          "opened for archive entry $pathname.");
    }

    var inputCompleter = new Completer();
    inputComplete = inputCompleter.future;

    _input = new ListInputStream();
    // TODO(nweiz): Report errors once issue 3657 is fixed
    var future = _consumeInput().chain((_) {
      if (!_input.closed) _input.markEndOfStream();
      // Asynchronously complete to give the InputStream callbacks a chance to
      // fire.
      return async();
    }).transform((_) => inputCompleter.complete(null));

    future.handleException((e) {
      print(e);
      print(future.stackTrace);
    });
    return _input;
  }

  /**
   * Close this entry so that its input stream no longer produces data.
   *
   * In addition to closing the associated input stream, this will prevent new
   * input streams from being opened.
   */
  void close() {
    _archiveId = null;
    if (_input != null) _input.close();
  }

  /**
   * Read all data from the archive and write it to [_input]. Returns a future
   * that completes once this is done.
   *
   * This assumes that both [_input] and [_archiveId] are non-null and that
   * [_input] is open, although if that changes before this completes it will
   * handle it gracefully.
   */
  Future _consumeInput() {
    var data;
    return call(read.DATA_BLOCK, _archiveId).chain((_data) {
      data = _data;
      // TODO(nweiz): This async() call is only necessary because of issue 4222.
      return async();
    }).chain((_) {
      if (_input.closed || _archiveId == null || data == null) {
        return new Future.immediate(null);
      }
      _input.write(data);
      return _consumeInput();
    });
  }
}

/**
 * An [ArchiveEntry] that contains the complete decompressed contents of the
 * file.
 */
class CompleteArchiveEntry extends ArchiveEntry {
  /** The contents of the entry as bytes. */
  final List<int> contentBytes;

  /** The contents of the entry as a string. */
  String get contents => new String.fromCharCodes(contentBytes);

  CompleteArchiveEntry._(List properties, this.contentBytes)
    : super.internal(properties, null);
}

File: utils/archive/entry.c
// Copyright (c) 2012, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

#include "entry.h"
#include "messaging.h"

#define DART_TIMESTAMP(name)                                      \
  Dart_CObject name;                                              \
  if (archive_entry_ ## name ## _is_set(e)) {                     \
    int64_t seconds = archive_entry_ ## name(e);                  \
    int64_t nanoseconds = archive_entry_ ## name ## _nsec(e);     \
    name.type = kInt64;                                           \
    name.value.as_int64 = seconds * 1000 + nanoseconds / 1000000; \
  } else {                                                        \
    name.type = kNull;                                            \
  }

#define RAW_ARCHIVE_SIZE 29

void postArchiveEntryArray(Dart_Port p, struct archive_entry* e) {
  if (checkPointerError(p, e, "archive entry")) return;

  Dart_CObject* archive_entry_array[RAW_ARCHIVE_SIZE];

  DART_INT64(id, (intptr_t) e);
  archive_entry_array[0] = &id;

  // archive_entry_paths(3)
  DART_STRING(hardlink, (char*) archive_entry_hardlink(e));
  archive_entry_array[1] = &hardlink;
  DART_STRING(pathname, (char*) archive_entry_pathname(e));
  archive_entry_array[2] = &pathname;
  DART_STRING(sourcepath, (char*) archive_entry_sourcepath(e));
  archive_entry_array[3] = &sourcepath;
  DART_STRING(symlink, (char*) archive_entry_symlink(e));
  archive_entry_array[4] = &symlink;

  // archive_entry_perms(3)
  DART_INT32(gid, archive_entry_gid(e));
  archive_entry_array[5] = &gid;
  DART_INT32(uid, archive_entry_uid(e));
  archive_entry_array[6] = &uid;
  DART_INT32(perm, archive_entry_perm(e));
  archive_entry_array[7] = &perm;
  DART_STRING(strmode, (char*) archive_entry_strmode(e));
  archive_entry_array[8] = &strmode;
  DART_STRING(gname, (char*) archive_entry_gname(e));
  archive_entry_array[9] = &gname;
  DART_STRING(uname, (char*) archive_entry_uname(e));
  archive_entry_array[10] = &uname;

  unsigned long fflags_set;
  unsigned long fflags_clear;
  archive_entry_fflags(e, &fflags_set, &fflags_clear);
  DART_INT64(wrapped_fflags_set, fflags_set);
  archive_entry_array[11] = &wrapped_fflags_set;
  DART_INT64(wrapped_fflags_clear, fflags_clear);
  archive_entry_array[12] = &wrapped_fflags_clear;

  DART_STRING(fflags_text, (char*) archive_entry_fflags_text(e));
  archive_entry_array[13] = &fflags_text;

  // archive_entry_stat(3)
  DART_INT32(filetype, archive_entry_filetype(e));
  archive_entry_array[14] = &filetype;
  DART_INT32(mode, archive_entry_mode(e));
  archive_entry_array[15] = &mode;

  Dart_CObject size;
  if (archive_entry_size_is_set(e)) {
    size.type = kInt64;
    size.value.as_int64 = archive_entry_size(e);
  } else {
    size.type = kNull;
  }
  archive_entry_array[16] = &size;

  Dart_CObject dev;
  if (archive_entry_dev_is_set(e)) {
    dev.type = kInt64;
    dev.value.as_int64 = archive_entry_dev(e);
  } else {
    dev.type = kNull;
  }
  archive_entry_array[17] = &dev;

  DART_INT64(devmajor, archive_entry_devmajor(e));
  archive_entry_array[18] = &devmajor;
  DART_INT64(devminor, archive_entry_devminor(e));
  archive_entry_array[19] = &devminor;

  Dart_CObject ino;
  if (archive_entry_ino_is_set(e)) {
    ino.type = kInt64;
    ino.value.as_int64 = archive_entry_ino64(e);
  } else {
    ino.type = kNull;
  }
  archive_entry_array[20] = &ino;

  DART_INT64(nlink, archive_entry_nlink(e));
  archive_entry_array[21] = &nlink;
  DART_INT64(rdev, archive_entry_rdev(e));
  archive_entry_array[22] = &rdev;
  DART_INT64(rdevmajor, archive_entry_rdevmajor(e));
  archive_entry_array[23] = &rdevmajor;
  DART_INT64(rdevminor, archive_entry_rdevminor(e));
  archive_entry_array[24] = &rdevminor;

  // archive_entry_time(3)
  DART_TIMESTAMP(atime);
  archive_entry_array[25] = &atime;
  DART_TIMESTAMP(birthtime);
  archive_entry_array[26] = &birthtime;
  DART_TIMESTAMP(ctime);
  archive_entry_array[27] = &ctime;
  DART_TIMESTAMP(mtime);
  archive_entry_array[28] = &mtime;
  // If you add entries, don't forget to increase RAW_ARCHIVE_SIZE.

  Dart_CObject wrapped_archive_entry;
  wrapped_archive_entry.type = kArray;
  wrapped_archive_entry.value.as_array.values = archive_entry_array;
  wrapped_archive_entry.value.as_array.length = RAW_ARCHIVE_SIZE;

  postSuccess(p, &wrapped_archive_entry);
}

void archiveEntryClone(Dart_Port p, struct archive_entry* e) {
  postArchiveEntryArray(p, archive_entry_clone(e));
}

void archiveEntryFree(Dart_Port p, struct archive_entry* e) {
  archive_entry_free(e);
  postSuccess(p, NULL);
}

void archiveEntryNew(Dart_Port p) {
  postArchiveEntryArray(p, archive_entry_new());
}

void archiveEntrySetHardlink(Dart_Port p, struct archive_entry* e,
    Dart_CObject* request) {
  Dart_CObject* value = getNullableStringArgument(p, request, 0);
  if (value == NULL) return;
  archive_entry_set_hardlink(e, getNullableString(value));
  postSuccess(p, NULL);
}

void archiveEntrySetPathname(Dart_Port p, struct archive_entry* e,
    Dart_CObject* request) {
  Dart_CObject* value = getNullableStringArgument(p, request, 0);
  if (value == NULL) return;
  archive_entry_set_pathname(e, getNullableString(value));
  postSuccess(p, NULL);
}

void archiveEntrySetSourcepath(Dart_Port p, struct archive_entry* e,
    Dart_CObject* request) {
  Dart_CObject* value = getNullableStringArgument(p, request, 0);
  if (value == NULL) return;
  archive_entry_set_sourcepath(e, getNullableString(value));
  postSuccess(p, NULL);
}

void archiveEntrySetSymlink(Dart_Port p, struct archive_entry* e,
    Dart_CObject* request) {
  Dart_CObject* value = getNullableStringArgument(p, request, 0);
  if (value == NULL) return;
  archive_entry_set_symlink(e, getNullableString(value));
  postSuccess(p, NULL);
}

void archiveEntrySetGid(Dart_Port p, struct archive_entry* e,
    Dart_CObject* request) {
  Dart_CObject* value = getIntArgument(p, request, 0);
  if (value == NULL) return;
  archive_entry_set_gid(e, getInteger(value));
  postSuccess(p, NULL);
}

void archiveEntrySetUid(Dart_Port p, struct archive_entry* e,
    Dart_CObject* request) {
  Dart_CObject* value = getIntArgument(p, request, 0);
  if (value == NULL) return;
  archive_entry_set_uid(e, getInteger(value));
  postSuccess(p, NULL);
}

void archiveEntrySetPerm(Dart_Port p, struct archive_entry* e,
    Dart_CObject* request) {
  Dart_CObject* value = getIntArgument(p, request, 0);
  if (value == NULL) return;
  archive_entry_set_perm_mask(e, getInteger(value));
  postSuccess(p, NULL);
}

void archiveEntrySetGname(Dart_Port p, struct archive_entry* e,
    Dart_CObject* request) {
  Dart_CObject* value = getNullableStringArgument(p, request, 0);
  if (value == NULL) return;
  archive_entry_update_gname_utf8(e, getNullableString(value));
  postSuccess(p, NULL);
}

void archiveEntrySetUname(Dart_Port p, struct archive_entry* e,
    Dart_CObject* request) {
  Dart_CObject* value = getNullableStringArgument(p, request, 0);
  if (value == NULL) return;
  archive_entry_update_uname_utf8(e, getNullableString(value));
  postSuccess(p, NULL);
}

void archiveEntrySetFflagsSet(Dart_Port p, struct archive_entry* e,
    Dart_CObject* request) {
  Dart_CObject* value = getIntArgument(p, request, 0);
  if (value == NULL) return;
  archive_entry_set_fflags(e, getInteger(value), 0);
  postSuccess(p, NULL);
}

void archiveEntrySetFflagsClear(Dart_Port p, struct archive_entry* e,
    Dart_CObject* request) {
  Dart_CObject* value = getIntArgument(p, request, 0);
  if (value == NULL) return;
  archive_entry_set_fflags(e, 0, getInteger(value));
  postSuccess(p, NULL);
}

void archiveEntrySetFflagsText(Dart_Port p, struct archive_entry* e,
    Dart_CObject* request) {
  Dart_CObject* value = getNullableStringArgument(p, request, 0);
  if (value == NULL) return;
  archive_entry_set_fflags_text(e, getNullableString(value));
  postSuccess(p, NULL);
}

void archiveEntrySetFiletype(Dart_Port p, struct archive_entry* e,
    Dart_CObject* request) {
  Dart_CObject* value = getIntArgument(p, request, 0);
  if (value == NULL) return;
  archive_entry_set_filetype(e, getInteger(value));
  postSuccess(p, NULL);
}

void archiveEntrySetMode(Dart_Port p, struct archive_entry* e,
    Dart_CObject* request) {
  Dart_CObject* value = getIntArgument(p, request, 0);
  if (value == NULL) return;
  archive_entry_set_mode(e, getInteger(value));
  postSuccess(p, NULL);
}

void archiveEntrySetSize(Dart_Port p, struct archive_entry* e,
    Dart_CObject* request) {
  if (request->type == kNull) {
    archive_entry_unset_size(e);
    postSuccess(p, NULL);
    return;
  }

  Dart_CObject* value = getIntArgument(p, request, 0);
  if (value == NULL) return;
  archive_entry_set_size(e, getInteger(value));
  postSuccess(p, NULL);
}

void archiveEntrySetDev(Dart_Port p, struct archive_entry* e,
    Dart_CObject* request) {
  Dart_CObject* value = getIntArgument(p, request, 0);
  if (value == NULL) return;
  archive_entry_set_dev(e, getInteger(value));
  postSuccess(p, NULL);
}

void archiveEntrySetDevmajor(Dart_Port p, struct archive_entry* e,
    Dart_CObject* request) {
  Dart_CObject* value = getIntArgument(p, request, 0);
  if (value == NULL) return;
  archive_entry_set_devmajor(e, getInteger(value));
  postSuccess(p, NULL);
}

void archiveEntrySetDevminor(Dart_Port p, struct archive_entry* e,
    Dart_CObject* request) {
  Dart_CObject* value = getIntArgument(p, request, 0);
  if (value == NULL) return;
  archive_entry_set_devminor(e, getInteger(value));
  postSuccess(p, NULL);
}

void archiveEntrySetIno(Dart_Port p, struct archive_entry* e,
    Dart_CObject* request) {
  Dart_CObject* value = getIntArgument(p, request, 0);
  if (value == NULL) return;
  archive_entry_set_ino64(e, getInteger(value));
  postSuccess(p, NULL);
}

void archiveEntrySetNlink(Dart_Port p, struct archive_entry* e,
    Dart_CObject* request) {
  Dart_CObject* value = getIntArgument(p, request, 0);
  if (value == NULL) return;
  archive_entry_set_nlink(e, getInteger(value));
  postSuccess(p, NULL);
}

void archiveEntrySetRdev(Dart_Port p, struct archive_entry* e,
    Dart_CObject* request) {
  Dart_CObject* value = getIntArgument(p, request, 0);
  if (value == NULL) return;
  archive_entry_set_rdev(e, getInteger(value));
  postSuccess(p, NULL);
}

void archiveEntrySetRdevmajor(Dart_Port p, struct archive_entry* e,
    Dart_CObject* request) {
  Dart_CObject* value = getIntArgument(p, request, 0);
  if (value == NULL) return;
  archive_entry_set_rdevmajor(e, getInteger(value));
  postSuccess(p, NULL);
}

void archiveEntrySetRdevminor(Dart_Port p, struct archive_entry* e,
    Dart_CObject* request) {
  Dart_CObject* value = getIntArgument(p, request, 0);
  if (value == NULL) return;
  archive_entry_set_rdevminor(e, getInteger(value));
  postSuccess(p, NULL);
}

void archiveEntrySetAtime(Dart_Port p, struct archive_entry* e,
    Dart_CObject* request) {
  if (request->type == kNull) {
    archive_entry_unset_atime(e);
    postSuccess(p, NULL);
    return;
  }

  Dart_CObject* value = getIntArgument(p, request, 0);
  if (value == NULL) return;
  int64_t atime = getInteger(value);
  archive_entry_set_atime(e, atime / 1000, (atime % 1000) * 1000000);
  postSuccess(p, NULL);
}

void archiveEntrySetBirthtime(Dart_Port p, struct archive_entry* e,
    Dart_CObject* request) {
  if (request->type == kNull) {
    archive_entry_unset_birthtime(e);
    postSuccess(p, NULL);
    return;
  }

  Dart_CObject* value = getIntArgument(p, request, 0);
  if (value == NULL) return;
  int64_t birthtime = getInteger(value);
  archive_entry_set_birthtime(
      e, birthtime / 1000, (birthtime % 1000) * 1000000);
  postSuccess(p, NULL);
}

void archiveEntrySetCtime(Dart_Port p, struct archive_entry* e,
    Dart_CObject* request) {
  if (request->type == kNull) {
    archive_entry_unset_ctime(e);
    postSuccess(p, NULL);
    return;
  }

  Dart_CObject* value = getIntArgument(p, request, 0);
  if (value == NULL) return;
  int64_t ctime = getInteger(value);
  archive_entry_set_ctime(e, ctime / 1000, (ctime % 1000) * 1000000);
  postSuccess(p, NULL);
}

void archiveEntrySetMtime(Dart_Port p, struct archive_entry* e,
    Dart_CObject* request) {
  if (request->type == kNull) {
    archive_entry_unset_mtime(e);
    postSuccess(p, NULL);
    return;
  }

  Dart_CObject* value = getIntArgument(p, request, 0);
  if (value == NULL) return;
  int64_t mtime = getInteger(value);
  archive_entry_set_mtime(e, mtime / 1000, (mtime % 1000) * 1000000);
  postSuccess(p, NULL);
}
